### PR TITLE
Track time spent in each wasm instance env call

### PR DESCRIPTION
**Pull request by kulakowski**
Additionally, this further tidies the interface between the host and the wasm instance env. In particular, reducer calls are explicitly started and stopped, and the wasm instance env becomes responsible for its own instrumentation.

# Description of Changes

# API and ABI

 - [ ] This is a breaking change to the module ABI
 - [ ] This is a breaking change to the module API
 - [ ] This is a breaking change to the ClientAPI
 - [ ] This is a breaking change to the SDK API

*If the API is breaking, please state below what will break*

# Expected complexity level and risk

2: This moderately changes some of the plumbing between the host and instance.

*How complicated do you think these changes are? Grade on a scale from 1 to 5,
where 1 is a trivial change, and 5 is a deep-reaching and complex change.*

*This complexity rating applies not only to the complexity apparent in the diff,
but also to its interactions with existing and future code.*

*If you answered more than a 2, explain what is complex about the PR,
and what other components it interacts with in potentially concerning ways.*

### CODE REVIEW

File: crates/core/src/host/wasm_common/instrumentation.rs [hunk 1]
```diff
@@ -0,0 +1,69 @@
+use std::time::{Duration, Instant};
```
Comment (kulakowski) [inline crates/core/src/host/wasm_common/instrumentation.rs:1]: The span stuff is rough and ready, happy to do whatever with it once we agree this is all the right direction.

File: crates/core/src/host/wasmer/wasm_instance_env.rs [hunk 2]
```diff
@@ -103,27 +120,57 @@ impl WasmInstanceEnv {
...
+        ExecutionTimings {
+            total_duration,
+            wasm_instance_env_call_times,
+        }
     }
 
     /// Call the function `f` with the name `func`.
     /// The function `f` is provided with the callers environment and the host's memory.
     ///
+    /// One of `cvt`, `cvt_ret`, or `cvt_noret` should be used in the implementation of any
```
Comment (kulakowski) [inline crates/core/src/host/wasmer/wasm_instance_env.rs:151]: What do you think of this structure?

**Review by gefjon — APPROVED** [review]
This looks good. I've left a few doc/comment related comments and suggestions, but I think the code itself is ready for the big time.

File: crates/core/src/host/wasm_common/instrumentation.rs [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+    /// Track a particular `CallSpan` by adding its duration to the
+    /// associated `Call`'s timing information.
+    pub fn span(&mut self, span: op::CallSpan) {
+        self.times[span.call] += span.duration;
+    }
+
+    /// Taking the record of call times gives a copy of the
+    /// current values and resets the values to zero.
+    pub fn take(&mut self) -> CallTimes {
```
[the remark this answers is not in the document]
Comment (kulakowski) [reply crates/core/src/host/wasm_common/instrumentation.rs:119]: I don't love the name, it roughly follows some other rust apis I've seen. I'll noodle on it.

[the remark this answers is not in the document]
Comment (gefjon) [reply crates/core/src/host/wasm_common/instrumentation.rs:119]: It's definitely the right name for this behavior, I just think it's not obvious why we need this behavior. Looking at the code, I've determined that we `take` the old timings after the end of a reducer run so we can report the ones we took and leave a fresh set of zeros in their place. I'd like to add to this docstring something like:

> `WasmInstanceEnv::finish_reducer` will `take` the `CallTimes` after running a reducer and report the taken times, leaving a fresh zeroed `CallTimes` for the next reducer invocation.

[the remark this answers is not in the document]
Comment (kulakowski) [reply crates/core/src/host/wasm_common/instrumentation.rs:120]: Done

**Pull request merged**